File: mixed/f1/MedianOfTwoSortedArrays.java
// Category: group3
// Given two sorted arrays, return the median of the two sorted arrays.
// The overall run time complexity should be O(log(m+n)), where m is the
// size of first and n is the size of the second array
// Sample: Input: nums1 = [1,4], nums2 = [3]
// Output: 2
// Explanation: merged array = [1,3,4] and the median is 3.
// Sample 2: Input: nums1 = [4,5], nums2 = [6,7]
// Output:
// Explanation: merged array = [4,5,6,7] and the median is (5+6)/2 = 5.5
// Soln: Brute Forrce - Merge and Sort

import java.util.Arrays;
class MedianOfTwoSortedArrays {
    public static void main(String...args) {
        int[] nums1 = {4,5};
        int[] nums2 = {6,7};
        System.out.println("Median: "+findMedianOfTwoSortedArrays(nums1,nums2));
    }

    static double findMedianOfTwoSortedArrays(int[] data1, int[] data2) {
        int m = data1.length;
        int n = data2.length;

        // Merge the two arrays into a single array
        int[] merged = new int[m+n];
        int k = 0;
        for(int i = 0; i < m; i++) {
            merged[k++] = data1[i];
        }
        for(int i = 0; i < n; i++) {
            merged[k++] = data2[i];
        }

        // Sort the merged array
        Arrays.sort(merged);

        int mergedLength = merged.length;

        if(mergedLength % 2 == 1) // odd number of elements: return the middle element as the median
            return (double) merged[mergedLength/2];
        else {
            // even number of elements: calculate the average of two middle elements as the median
            int middle1 = merged[mergedLength / 2 - 1];
            int middle2 = merged[mergedLength / 2];

            return ((double) middle1 + (double) middle2) / 2.0;
        }
    }
}

File: mixed/f1/median-of-two-sorted-arrays.go
// Category: group3
// Given two sorted arrays, return the median of the two sorted arrays.
// The overall run time complexity should be O(log(m+n)), where m is the
// size of first and n is the size of the second array
// Sample: Input: nums1 = [1,4], nums2 = [3]
// Output: 2
// Explanation: merged array = [1,3,4] and the median is 3.
// Sample 2: Input: nums1 = [4,5], nums2 = [6,7]
// Output:
// Explanation: merged array = [4,5,6,7] and the median is (5+6)/2 = 5.5
// Soln: Brute Force - Merge and Sort
package main

import(
    "fmt"
    "sort"
)

func find_median_of_two_sorted_arrays(data1 []int, data2 []int) float32 {
    merged := append(data1, data2...)
    merged_length := len(merged)
    sort.Ints(merged)

    if merged_length % 2 == 1 {
        return (float32)(merged[merged_length / 2])
    } else {
        middle1 := merged[merged_length / 2 - 1]
        middle2 := merged[merged_length / 2]
        return ((float32)(middle1) + (float32)(middle2)) / 2
    }
}

func main() {
    nums1 := []int{4, 5}
    nums2 := []int{6, 7}
    fmt.Printf("Median: %.2f\n", find_median_of_two_sorted_arrays(nums1, nums2))
}